rollup/rollupsyncservice/rollupsyncservice.go

```diff
@@ -0,0 +1,395 @@
+	for _, vLog := range logs {
+		switch vLog.Topics[0] {
+		case s.l1CommitBatchEventSignature:
+			event := &L1CommitBatchEvent{}
+			if err := UnpackLog(s.scrollChainABI, event, "CommitBatch", vLog); err != nil {
+				return fmt.Errorf("failed to unpack commit rollup event log, err: %w", err)
+			}
+			batchIndex := event.BatchIndex.Uint64()
+
+			chunkBlockRanges, err := s.getChunkRanges(batchIndex, &vLog)
+			if err != nil {
+				return fmt.Errorf("failed to get chunk ranges, err: %w", err)
+			}
+			rawdb.WriteBatchChunkRanges(s.db, batchIndex, chunkBlockRanges)
+
+		case s.l1RevertBatchEventSignature:
+			event := &L1RevertBatchEvent{}
+			if err := UnpackLog(s.scrollChainABI, event, "RevertBatch", vLog); err != nil {
+				return fmt.Errorf("failed to unpack revert rollup event log, err: %w", err)
+			}
+			batchIndex := event.BatchIndex.Uint64()
+
+			rawdb.DeleteBatchChunkRanges(s.db, batchIndex)
+
+		case s.l1FinalizeBatchEventSignature:
+			event := &L1FinalizeBatchEvent{}
+			if err := UnpackLog(s.scrollChainABI, event, "FinalizeBatch", vLog); err != nil {
+				return fmt.Errorf("failed to unpack finalized rollup event log, err: %w", err)
+			}
+			batchIndex := event.BatchIndex.Uint64()
+
+			parentBatchMeta, chunks, err := s.getLocalInfoForBatch(batchIndex)
+			if err != nil {
+				return fmt.Errorf("failed to get local node info, batch index: %v, err: %w", batchIndex, err)
+			}
+
+			if err := validateBatch(event, parentBatchMeta, chunks, s.node); err != nil {
+				return fmt.Errorf("fatal: validateBatch failed: finalize event: %v, err: %w", *event, err)
+			}
+			endChunk := chunks[len(chunks)-1]
+			endBlock := endChunk.Blocks[len(endChunk.Blocks)-1]
+
+			rawdb.WriteFinalizedL2BlockNumber(s.db, endBlock.Header.Number.Uint64())
+			rawdb.WriteFinalizedBatchMeta(s.db, batchIndex, calculateFinalizedBatchMeta(parentBatchMeta, event.BatchHash, chunks))
+
+			if batchIndex%100 == 0 {
+				log.Info("finalized batch progress", "batch index", batchIndex, "finalized l2 block height", endBlock.Header.Number.Uint64())
+			}
+
+		default:
+			return fmt.Errorf("unknown event, topic: %v, tx hash: %v", vLog.Topics[0].Hex(), vLog.TxHash.Hex())
+		}
+	}
+
+	// note: the batch updates above are idempotent, if we crash
+	// before this line and reexecute the previous steps, we will
+	// get the same result.
+	rawdb.WriteRollupEventSyncedL1BlockNumber(s.db, endBlockNumber)
+	s.latestProcessedBlock = endBlockNumber
+
+	return nil
+}
+
+func (s *RollupSyncService) getLocalInfoForBatch(batchIndex uint64) (*rawdb.FinalizedBatchMeta, []*Chunk, error) {
+	chunkBlockRanges := rawdb.ReadBatchChunkRanges(s.db, batchIndex)
+	if len(chunkBlockRanges) == 0 {
+		return nil, nil, fmt.Errorf("failed to get batch chunk ranges, empty chunk block ranges")
+	}
+
+	endBlockNumber := chunkBlockRanges[len(chunkBlockRanges)-1].EndBlockNumber
+	for i := 0; i < defaultMaxRetries; i++ {
```

Comment:
Should we exit the loop early in case of `ctx.Done`?

Reply:
nice catch, added in [ac7bdb3](https://github.com/scroll-tech/go-ethereum/pull/515/commits/ac7bdb393f54b1b7974a6794734992db2813061c)